File: asd-millennio-backend/src/schemas/calendario.py
from datetime import date, datetime, time
from uuid import UUID

from pydantic import BaseModel


class SlotResponse(BaseModel):
    id: UUID
    data: date
    fascia: str
    ora_inizio: time
    ora_fine: time
    ore_totali: int
    ore_vendute: list[int]
    ore_in_lock: list[int]
    stato: str
    nft_token_id: int | None
    bloccato_fino_a: datetime | None

    model_config = {"from_attributes": True}


class DisponibilitaFascia(BaseModel):
    slot_id: UUID
    data: date
    fascia: str
    ora_inizio: time
    ora_fine: time
    ore_totali: int
    ore_vendute: list[int]
    ore_in_lock: list[int]
    ore_libere: list[int]
    stato: str
    prezzo_ora: float
    moltiplicatore_data: float
    moltiplicatore_scarsita: float
    sconto_promo_pct: float


class OreSelezione(BaseModel):
    slot_id: UUID
    ore: list[int]


class LockRequest(BaseModel):
    selezione: list[OreSelezione]


class DettaglioSlot(BaseModel):
    slot_id: str
    fascia: str
    data: str
    ore_selezionate: list[int]
    n_ore: int
    tariffa_base: float
    moltiplicatore_data: float
    moltiplicatore_scarsita: float
    sconto_promo_pct: float
    prezzo_ora: float
    costo_slot: float


class RiepilogoSelezione(BaseModel):
    selezione: list[dict]
    ore_notte: int
    ore_mattina: int
    ore_pomeriggio: int
    costo_totale: float
    dettaglio: list[DettaglioSlot] = []

File: asd-millennio-frontend/lib/api/calendario.ts
import { apiClient } from "./client";

export interface SlotDisponibilita {
  slot_id: string;
  data: string;           // YYYY-MM-DD
  fascia: "notte" | "mattina" | "pomeriggio";
  ora_inizio: string;     // HH:MM:SS
  ora_fine: string;
  ore_totali: number;
  ore_vendute: number[];  // es. [0,1,2] = quelle ore già vendute
  ore_in_lock: number[];  // ore in lock da altri utenti
  ore_libere: number[];   // ore ancora selezionabili
  stato: "libero" | "parziale" | "esaurito" | "bloccato";
  prezzo_ora: number;
  moltiplicatore_data: number;
  moltiplicatore_scarsita: number;
  sconto_promo_pct: number;
}

export interface OreSelezione {
  slot_id: string;
  ore: number[];
}

export interface DettaglioSlot {
  slot_id: string;
  fascia: string;
  data: string;
  ore_selezionate: number[];
  n_ore: number;
  tariffa_base: number;
  moltiplicatore_data: number;
  moltiplicatore_scarsita: number;
  sconto_promo_pct: number;
  prezzo_ora: number;
  costo_slot: number;
}

export interface RiepilogoSelezione {
  selezione: OreSelezione[];
  ore_notte: number;
  ore_mattina: number;
  ore_pomeriggio: number;
  costo_totale: number;
  dettaglio: DettaglioSlot[];
}

export async function fetchDisponibilita(params: {
  data_inizio?: string;
  data_fine?: string;
}): Promise<SlotDisponibilita[]> {
  const { data } = await apiClient.get<SlotDisponibilita[]>("/calendario/disponibilita", { params });
  return data;
}

export async function lockSelezione(selezione: OreSelezione[]): Promise<RiepilogoSelezione> {
  const { data } = await apiClient.post<RiepilogoSelezione>("/calendario/lock", { selezione });
  return data;
}
